pkg/splunk/resources/util.go

```diff
@@ -197,8 +198,86 @@ func CompareEnvs(a []corev1.EnvVar, b []corev1.EnvVar) bool {
 	return false
 }
 
-// SortVolumeMounts returns a sorted list of Kubernetes VolumeMounts.
-func SortVolumeMounts(mounts []corev1.VolumeMount) []corev1.VolumeMount {
+// sortVolumes returns a sorted list of Kubernetes Volumes.
+func sortVolumes(volumes []corev1.Volume) []corev1.Volume {
+	sorted := make([]corev1.Volume, len(volumes))
+	copy(sorted, volumes)
+	sort.Slice(sorted, func(i, j int) bool { return sorted[i].Name < sorted[j].Name })
+	return sorted
+}
+
+func compareSecretVolume(a, b *corev1.SecretVolumeSource) bool {
+	if a.SecretName != b.SecretName {
+		return true
+	}
+
+	if a.DefaultMode != nil && b.DefaultMode != nil {
+		if *a.DefaultMode != *b.DefaultMode {
+			return true
+		}
+	}
+	return false
+}
+
+func compareConfigMapVolume(a, b *corev1.ConfigMapVolumeSource) bool {
+	if a.Name != b.Name {
+		return true
+	}
+
+	if a.DefaultMode != nil && b.DefaultMode != nil {
```

Comment:
Assuming these comparison methods are necessary, I think these pointer comparisons need to be updated to account for one being nil while the other is not.

Reply:
k8s always initializes this field per https://github.com/kubernetes/kubernetes/blob/94837daab20ce5b3fee00f26483a4d64c4d7b9e4/pkg/apis/core/v1/defaults.go#L237 . So if one is nil (ie "spec"), there is no need to check for change.

Reply:
If we make the assumption that this is always only used to compare K8s-updated version to one generated by operator, it effectively means it is dead code because we know b.DefaultMode is always nil. In this case, I think it would be better to just not have anything. My bigger concern is that these aren't really performing the comparisons one would expect them to be performing, which means there is a high likelihood they would eventually be misused in such a way that introduces bugs.

Reply:
Not sure I followed. The k8s-updated version always set DefaultMode to 420 if nil. But the operator version can be nil or non-nil (ie user can specified it in the CR).